### PR TITLE
[Merged by Bors] - Move magic numbers into constants in Breakout example

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- The Breakout example has a lot of configurable constant values for setup, but these are buried in the source code.
- Magic numbers scattered in the source code are hard to follow.
- Providing constants up front makes tweaking examples very approachable.

## Solution

- Move magic numbers into constants

## Context

Part of the changes made in #2094; split out for easier review.

### CODE REVIEW

File: examples/game/breakout.rs [hunk 1]
```diff
@@ -1,16 +1,54 @@
...
     sprite::collide_aabb::{collide, Collision},
 };
 
-/// An implementation of the classic game "Breakout"
+// Sets the amount of time that should elapse between each physics step
 const TIME_STEP: f32 = 1.0 / 60.0;
+
+// These constants are in `Transform` units:
+// with the default 2D camera they correspond 1:1 with screen pixels
+// These macros are needed as functions that operate on floats cannot be constant in Rust :(
```
Comment (hymm) [inline examples/game/breakout.rs:15]:
```suggestion
// The `const_vec3!` macros are needed as functions that operate on floats cannot be constant in Rust.
```
A little more specific and remove the frowny face as it feels like you're disparaging rust, rather than a cheeky comment.

File: examples/game/breakout.rs [hunk 2]
```diff
@@ -1,16 +1,54 @@
...
+
+const BRICK_ROWS: isize = 4;
+const BRICK_COLUMNS: isize = 5;
+const BRICK_SPACING: f32 = 20.0;
+const BRICK_SIZE: Vec3 = const_vec3!([150.0, 30.0, 1.0]);
+
+const SCOREBOARD_FONT_SIZE: f32 = 40.0;
+const SCOREBOARD_TEXT_PADDING: Val = Val::Px(5.0);
+
+const BACKGROUND_COLOR: Color = Color::rgb(0.95, 0.95, 0.95);
```
Comment (hymm) [inline examples/game/breakout.rs:39]: These colors choices feel worse to me than the original. I don't think we can use the built in colors here as they tend to be too saturated. Maybe try taking a palete from something like https://www.colourlovers.com/palettes or use a color picker like https://color.adobe.com/create/color-wheel

File: examples/game/breakout.rs [hunk 3]
```diff
@@ -1,16 +1,54 @@
...
+const PADDLE_HEIGHT: f32 = -215.0;
+const PADDLE_SIZE: Vec3 = const_vec3!([120.0, 30.0, 0.0]);
+const PADDLE_SPEED: f32 = 500.0;
+const PADDLE_BOUNDS: f32 = 380.0;
+
+// We set the z-value of the ball to 1 so it renders on top in the case of overlapping sprites
+const BALL_STARTING_POSITION: Vec3 = const_vec3!([0.0, -50.0, 1.0]);
+const BALL_SIZE: Vec3 = const_vec3!([30.0, 30.0, 0.0]);
+const BALL_SPEED: f32 = 400.0;
+// This vector is later extended into three dimensions and then normalized
```
Comment (hymm) [inline examples/game/breakout.rs:25]: this comment feels extraneous. Especially with how disconnected it is from where it actually is done.

File: examples/game/breakout.rs [hunk 4]
```diff
@@ -1,16 +1,54 @@
...
+const INITIAL_BALL_DIRECTION: Vec2 = const_vec2!([0.5, -0.5]);
+
+const PLAY_AREA_BOUNDS: Vec2 = const_vec2!([900.0, 600.0]);
+const WALL_THICKNESS: f32 = 10.0;
+
+const BRICK_ROWS: isize = 4;
```
Comment (hymm) [inline examples/game/breakout.rs:31]: Should these be usize or u32 since they can't be negative? Probably u32 since they don't really act as indicies?

**Review by hymm — COMMENTED** [review]
Overall I think moving the relevant constants to the top of the file makes sense as it does let beginners play with things more easily.

File: examples/game/breakout.rs (hunk 3 above)
Comment (alice-i-cecile) [reply examples/game/breakout.rs:25]:
```suggestion
```

File: examples/game/breakout.rs (hunk 4 above)
Comment (alice-i-cecile) [reply examples/game/breakout.rs:30]: They must be i something, as they're strictly non-negative.

File: examples/game/breakout.rs (hunk 2 above)
Comment (alice-i-cecile) [reply examples/game/breakout.rs:38]: My preference was for "ease of reading the code" over "pretty". IMO it's important to use built-in constants here, because the name lets readers connect what they see to the code more easily.

File: examples/game/breakout.rs (hunk 4 above)
Comment (hymm) [reply examples/game/breakout.rs]: u = unsigned

File: examples/game/breakout.rs [hunk 5]
```diff
@@ -1,16 +1,53 @@
...
+// We set the z-value of the ball to 1 so it renders on top in the case of overlapping sprites
+const BALL_STARTING_POSITION: Vec3 = const_vec3!([0.0, -50.0, 1.0]);
+const BALL_SIZE: Vec3 = const_vec3!([30.0, 30.0, 0.0]);
+const BALL_SPEED: f32 = 400.0;
+const INITIAL_BALL_DIRECTION: Vec2 = const_vec2!([0.5, -0.5]);
+
+const PLAY_AREA_BOUNDS: Vec2 = const_vec2!([900.0, 600.0]);
+const WALL_THICKNESS: f32 = 10.0;
+
+const BRICK_ROWS: i8 = 4;
```
Comment (hymm) [inline examples/game/breakout.rs:30]: u mean unsigned so should probably be a u8.

**Review by hymm — APPROVED** [review]

File: examples/game/breakout.rs [hunk 6]
```diff
@@ -1,16 +1,53 @@
...
     prelude::*,
     sprite::collide_aabb::{collide, Collision},
 };
 
-/// An implementation of the classic game "Breakout"
+// Sets the amount of time that should elapse between each physics step
 const TIME_STEP: f32 = 1.0 / 60.0;
+
+// These constants are in `Transform` units:
+// with the default 2D camera they correspond 1:1 with screen pixels
```
Comment (KDecay) [inline examples/game/breakout.rs:14]:
```suggestion
// These constants are defined in `Transform` units.
// Using the default 2D camera they correspond 1:1 with screen pixels.
```

File: examples/game/breakout.rs [hunk 7]
```diff
@@ -1,16 +1,53 @@
...
+
 use bevy::{
     core::FixedTimestep,
+    math::{const_vec2, const_vec3},
     prelude::*,
     sprite::collide_aabb::{collide, Collision},
 };
 
-/// An implementation of the classic game "Breakout"
+// Sets the amount of time that should elapse between each physics step
```
Comment (KDecay) [inline examples/game/breakout.rs:10]:
```suggestion
// Defines the amount of time that should elapse between each physics step.
```

File: examples/game/breakout.rs [hunk 8]
```diff
@@ -1,16 +1,53 @@
...
+
+// These constants are in `Transform` units:
+// with the default 2D camera they correspond 1:1 with screen pixels
+// The `const_vec3!` macros are needed as functions that operate on floats cannot be constant in Rust.
+const PADDLE_HEIGHT: f32 = -215.0;
+const PADDLE_SIZE: Vec3 = const_vec3!([120.0, 30.0, 0.0]);
+const PADDLE_SPEED: f32 = 500.0;
+const PADDLE_BOUNDS: f32 = 380.0;
+
+// We set the z-value of the ball to 1 so it renders on top in the case of overlapping sprites
```
Comment (KDecay) [inline examples/game/breakout.rs:21]:
```suggestion
// We set the z-value of the ball to 1 so it renders on top in the case of overlapping sprites.
```

File: examples/game/breakout.rs [hunk 9]
```diff
@@ -55,34 +92,37 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
...
             sprite: Sprite {
-                color: Color::rgb(1.0, 0.5, 0.5),
+                color: BALL_COLOR,
                 ..default()
             },
             ..default()
         })
         .insert(Ball {
-            velocity: 400.0 * Vec3::new(0.5, -0.5, 0.0).normalize(),
+            // We can create a velocity by multiplying our speed by a normalized direction
```
Comment (KDecay) [inline examples/game/breakout.rs:124]:
```suggestion
            // We can create a velocity by multiplying our speed by a normalized direction.
```

**Review by KDecay — CHANGES_REQUESTED** [review]
LGTM. As a beginner I would've greatly appreciated the constants so this change makes sense to me. 

I think the amount of documentation on the constants is fine as is for now, because otherwise the example would be drowned in comments that don't add a lot to the example itself. 

If we see users asking questions about particular constants we should add documentation to those, but they are mostly self explanatory.

**Review by KDecay — APPROVED** [review]

File: examples/game/breakout.rs [hunk 10]
```diff
@@ -1,16 +1,53 @@
...
 };
 
-/// An implementation of the classic game "Breakout"
+// Defines the amount of time that should elapse between each physics step.
 const TIME_STEP: f32 = 1.0 / 60.0;
+
+// These constants are defined in `Transform` units.
+// Using the default 2D camera they correspond 1:1 with screen pixels.
+// The `const_vec3!` macros are needed as functions that operate on floats cannot be constant in Rust.
+const PADDLE_HEIGHT: f32 = -215.0;
```
Comment (mockersf) [inline examples/game/breakout.rs:16]: the meaning of this one is not very obvious, even less with the `PADDLE_SIZE` just under

Comment (alice-i-cecile) [reply examples/game/breakout.rs:16]: Agreed. I'll fix this in a follow-up PR.
